src/helpers/privacy_budget.rs

```diff
@@ -0,0 +1,217 @@
+use crate::error::{Error, Res};
+use crate::threshold::Ciphertext;
+use hex::encode as hex;
+/// This module implements privacy budget using a redis cache
+use redis::{transaction, Commands, RedisError};
+use rust_elgamal::{DecryptionKey, RistrettoPoint, Scalar};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct BlindedMatchKey(Ciphertext);
+
+#[allow(dead_code)]
+impl BlindedMatchKey {
+    pub fn new(key: DecryptionKey, message: RistrettoPoint, scalar: Scalar) -> Self {
+        let c = key.encryption_key().encrypt_with(message, scalar);
+        Self(c)
+    }
+}
+
+impl ToString for BlindedMatchKey {
+    fn to_string(&self) -> String {
+        let (c0, c1) = self.0.inner();
+        let mut c0_hex = hex(c0.compress().to_bytes());
+        let c1_hex = hex(c1.compress().to_bytes());
+        c0_hex.push_str(&c1_hex);
+        c0_hex
+    }
+}
+
+struct PrivacyBudget {
+    redis_connection: redis::Connection,
+    total_privacy_budget: u32,
+}
+
+trait ToString {
+    fn to_string(&self) -> String;
+}
+
+impl ToString for Ciphertext {
+    fn to_string(&self) -> String {
+        let (c0, c1) = self.inner();
+        let mut c0_hex = hex(c0.compress().to_bytes());
+        let c1_hex = hex(c1.compress().to_bytes());
+        c0_hex.push_str(&c1_hex);
+        c0_hex
+    }
+}
+
+#[allow(dead_code)]
+impl PrivacyBudget {
+    fn new(
+        redis_host_name: &str,
+        redis_password: Option<String>,
```

Comment:
&str would seem to be sufficient for password (if you have no password, then the empty string can be provided).